networkx/drawing/nx_pylab.py

```diff
@@ -683,122 +681,155 @@ def draw_networkx_edges(
-                data_loc + np.asarray([h_shift, 0]),
-                data_loc,
-                # 4 4 4
-                data_loc + np.asarray([-h_shift, 0]),
-                data_loc + np.asarray([-h_shift, v_shift]),
-                data_loc + np.asarray([0, v_shift]),
-            ]
-
-            ret = mpl.path.Path(ax.transData.transform(path), [1, 4, 4, 4, 4, 4, 4])
-        # if not, fall back to the user specified behavior
-        else:
-            ret = base_connection_style(posA, posB, *args, **kwargs)
-
-        return ret
-
-    # FancyArrowPatch doesn't handle color strings
-    arrow_colors = mpl.colors.colorConverter.to_rgba_array(edge_color, alpha)
-    for i, (src, dst) in enumerate(edge_pos):
-        x1, y1 = src
-        x2, y2 = dst
-        shrink_source = 0  # space from source to tail
-        shrink_target = 0  # space from  head to target
-        if np.iterable(node_size):  # many node sizes
-            source, target = edgelist[i][:2]
-            source_node_size = node_size[nodelist.index(source)]
-            target_node_size = node_size[nodelist.index(target)]
-            shrink_source = to_marker_edge(source_node_size, node_shape)
-            shrink_target = to_marker_edge(target_node_size, node_shape)
-        else:
-            shrink_source = shrink_target = to_marker_edge(node_size, node_shape)
-
-        if shrink_source < min_source_margin:
-            shrink_source = min_source_margin
-
-        if shrink_target < min_target_margin:
-            shrink_target = min_target_margin
-
-        if len(arrow_colors) == len(edge_pos):
-            arrow_color = arrow_colors[i]
-        elif len(arrow_colors) == 1:
-            arrow_color = arrow_colors[0]
-        else:  # Cycle through colors
-            arrow_color = arrow_colors[i % len(arrow_colors)]
-
-        if np.iterable(width):
-            if len(width) == len(edge_pos):
-                line_width = width[i]
-            else:
-                line_width = width[i % len(width)]
-        else:
-            line_width = width
-
-        arrow = mpl.patches.FancyArrowPatch(
-            (x1, y1),
-            (x2, y2),
-            arrowstyle=arrowstyle,
-            shrinkA=shrink_source,
-            shrinkB=shrink_target,
-            mutation_scale=mutation_scale,
-            color=arrow_color,
-            linewidth=line_width,
-            connectionstyle=_connectionstyle,
-            linestyle=style,
-            zorder=1,
-        )  # arrows go behind nodes
-
-        arrow_collection.append(arrow)
-        ax.add_patch(arrow)
+    # Draw the edges
+    if use_linecollection:
+        edge_viz_obj = _draw_networkx_edges_line_collection()
+        # Make sure selfloop edges are also drawn.
+        edgelist = list(nx.selfloop_edges(G))
+        if edgelist:
+            edge_pos = np.asarray([(pos[e[0]], pos[e[1]]) for e in edgelist])
+            arrowstyle = "-"
+            _draw_networkx_edges_fancy_arrow_patch()
```

Comment:
Should the FancyArrowPatch arrow_collection created for self-loops with undirected graphs (which is just a list if I read correctly)  get returned somehow?
I don't think it can be easily combined with LineCollection...  but I don't know much about combining the parts of a matplotlib drawing. It **does** get added to `ax` as a patch. So maybe that is enough.

I imagine not very many people will need to access those self-loops directly, but we provide that access for the other edges. Can we find a way to do it for these?

Reply:
I had thought about this as well - one of the initial issues was about how different objects get returned based on the the inputs... We could in principle homogenize the output so that it always returns a tuple, e.g. `return (LineCollection, FancyArrowPatches)` where one or the other inputs may be empty depending on whether they were used in the drawing or not. If this were a brand new function, we could consider that; but since this change would break backward compatibility I don't think it's a good idea.

IMO the best way to access the edges to modify them after they've been created by `draw_networkx_edges` is through the `Axes` attributes like you've pointed out, e.g.

```python
ax = plt.gca()
ax.patches  # FancyArrowPatches for directed edges would can be accessed in this list
ax.collections  # LineCollections for undirected edges can be accessed here.
```

My preference would be to leave the current return types as-is and maybe just add a note to the documentation about how to access the various objects representing various types of edges.